### PR TITLE
11 get/api/articles queries

### PR DESCRIPTION
**Pull request by woodsjd11**
- refactored GET /api/articles endpoint
- add sort_by query
- add order query
- add topic query
- protected against sql injection

### CODE REVIEW

File: models/index.js [hunk 1]
```diff
@@ -23,14 +24,39 @@ exports.fetchUsers = () => {
...
+    topicFilter.push(topic);
+  }
+  queryStr += ` GROUP BY articles.article_id ORDER BY ${sort_by} ${order}`;
+  
+  return db.query(queryStr, topicFilter).then(({ rows, rowCount }) => {
+    if (rowCount > 0) {
       return rows;
-    });
+    }
+    return Promise.reject({ status: 404, message: "Topic Not Found" });
```
Comment (two-triangles) [inline models/index.js:58]: It isn't too much of an issue now, - but one of the extra tasks later on is to post a topic. This initially may have no articles so the rowCount would be 0. The current error message would conflict with that. There is not anything wrong with sending back an empty in this case, as a topic that doesn't exist yet - may exist in the future

**Comment by two-triangles** [comment]
just the one comment, I would refactor this to an empty array. Once this is done let me know, and it should be good for a merge!

File: __tests__/app.test.js [hunk 2]
```diff
@@ -275,6 +294,14 @@ describe("GET /api/articles", () => {
           });
         });
     });
+    test("200: currently the topic does not exist, returns empty", () => {
+      return request(app)
+        .get("/api/articles?sort_by=author&order=asc&topic=buns")
+        .expect(200)
+        .then(({ body: { articles } }) => {
+          expect(articles).toHaveLength(0);
```
Comment (emilyb93) [inline __tests__/app.test.js:302]: We need to check the data type too as length can be a property on a string also. We can either check the type or check that it equals []

File: models/index.js [hunk 3]
```diff
@@ -23,14 +24,39 @@ exports.fetchUsers = () => {
...
+    topicValue.push(topic);
+  }
+  queryStr += ` GROUP BY articles.article_id ORDER BY ${sort_by} ${order}`;
+  
+  return db.query(queryStr, topicValue).then(({ rows, rowCount }) => {
+    // if (rowCount > 0) {
       return rows;
-    });
+    // }
+    // return Promise.reject({ status: 404, message: "Topic Not Found" });
```
Comment (emilyb93) [inline models/index.js:58]: remove comments if your code is working. Comments are for explaining your code not for indecision

File: test.sql [hunk 4]
```diff
@@ -1,3 +1,3 @@
 \c nc_news_test;
 
-SELECT * FROM users WHERE username = "wrong"
\ No newline at end of file
+SELECT articles.author, title, articles.article_id, topic, articles.created_at, articles.votes, COUNT(comments.article_id)::int AS comment_count FROM articles LEFT JOIN comments ON comments.article_id = articles.article_id GROUP BY articles.article_id ORDER BY articles.created_at DESC
```
Comment (emilyb93) [inline test.sql:3]: dont commit files that are not relevant

File: test.txt [hunk 5]
```diff
@@ -1,6 +1,17 @@
...
+ rogersop      | UNCOVERED: catspiracy to bring down democracy          |          5 | cats  | 2020-08-03 14:14:00 |     0 |             2
+ butter_bridge | Living in the shadow of a great man                    |          1 | mitch | 2020-07-09 21:11:00 |   100 |            11
+ butter_bridge | They're not exactly dogs, are they?                    |          9 | mitch | 2020-06-06 10:10:00 |     0 |             2
+ rogersop      | Seven inspirational thought leaders from Manchester UK |         10 | mitch | 2020-05-14 05:15:00 |     0 |             0
+ rogersop      | Student SUES Mitch!                                    |          4 | mitch | 2020-05-06 02:14:00 |     0 |             0
+ icellusedkars | Does Mitch predate civilisation?                       |          8 | mitch | 2020-04-17 02:08:00 |     0 |             0
+ icellusedkars | Am I a cat?                                            |         11 | mitch | 2020-01-15 22:21:00 |     0 |             0
+ icellusedkars | Z                                                      |          7 | mitch | 2020-01-07 14:08:00 |     0 |             0
+(12 rows)
```
Comment (emilyb93) [inline test.txt:16]: dont commit files that arent relevant

**Review by emilyb93 — APPROVED** [review]
Remove comments and unnecessary files

File: __tests__/app.test.js [hunk 6]
```diff
@@ -232,47 +258,79 @@ describe("GET /api/articles", () => {
...
+      return request(app)
+        .get("/api/articles")
+        .expect(200)
+        .then(({ body: { articles } }) => {
+          expect(articles).toBeSortedBy("created_at", { descending: true });
+        });
+    });
+    test("200: returns articles sorted by any valid column, ordered by default descending", () => {
+      return request(app)
+        .get("/api/articles?sort_by=title")
```
Comment (two-triangles) [inline __tests__/app.test.js:272]: you can. make use of the .query method to pass in your queries as well

File: __tests__/app.test.js [hunk 7]
```diff
@@ -232,47 +258,79 @@ describe("GET /api/articles", () => {
...
+          expect(articles).toBeSortedBy("author", { ascending: true });
+        });
+    });
+    test("200: returns articles filtered by topic", () => {
+      return request(app)
+        .get("/api/articles?sort_by=author&order=asc&topic=mitch")
+        .expect(200)
+        .then(({ body: { articles } }) => {
+          expect(articles).toBeSortedBy("author", { ascending: true });
+          articles.forEach((article) => {
```
Comment (two-triangles) [inline __tests__/app.test.js:292]: If we know that there will be multiple artucles of the same topic, we should be making an assertion on how many we expect to get back

File: __tests__/app.test.js [hunk 8]
```diff
@@ -232,47 +258,79 @@ describe("GET /api/articles", () => {
...
+            expect(article.topic).toBe("mitch");
+          });
+        });
+    });
+    test("200: valid topic, but topic does not exist on article, hence returns empty", () => {
```
Comment (two-triangles) [inline __tests__/app.test.js:297]: empty array?

**Pull request merged**